Transform scala dependencies (#78)
r2c-CSE/semgrep-utilities: README.md

bash.lang.security.ifs-tampering.ifs-tampering,0.044
````

## Scala Transformation

Scala is currently not on our list of [supported languages](https://semgrep.dev/docs/supported-languages/#semgrep-supply-chain) for SSC. However, Scala is a Java derivative. And while Java rules can attend the Scala party, reachability rules are like that one friend who didn't get the invite. So for now, Scala's flying solo with parity and lockfile support, no reachability plus-one. That being said, we don’t have any specific support for Scala ecosystems/package files.

> A few customers use SBT to construct their dependency tree. For those unfamiliar, a guide to use SBT can be found [here](https://www.baeldung.com/scala/sbt-dependency-tree).

Reviewer: ```suggestion
If you use SBT to construct your dependency tree, this script can help you transform it for Semgrep Supply Chain to scan. For those unfamiliar, a guide to use SBT can be found [here](https://www.baeldung.com/scala/sbt-dependency-tree).
```

### SBT Dependency Tree Output

Using the **`sbt dependencyTree`** command yields a dependency tree that is SBT-specific but akin to the one required by SSC.

> To run this command, the user first needs to add the appropriate plugin:

Reviewer: ```suggestion
To run this command, first add the appropriate plugin:
```

**`addDependencyTreePlugin`** 

Or for sbt ≤ 1.3, 
`addSbtPlugin("net.virtual-void" % "sbt-dependency-graph" % "0.10.0-RC1")` 

(see https://github.com/sbt/sbt-dependency-graph)

If any of your Scala projects have deeply nested dependencies, you probably will also want to run `sbt 'set asciiGraphWidth := 9999'` . This prevents SBT from truncating the dependency information, which will most likely lead to errors in the later transform step if not addressed.

### Transform script

> SSC scans require a **`maven_dep_tree.txt`** file in a format which is generated by the command mentioned [here](https://semgrep.dev/docs/semgrep-supply-chain/setup-maven/). To bridge the gap between the SBT dep tree and the Maven format, we approached the transformation using two methods:

Reviewer: ```suggestion
SSC scans require a **`maven_dep_tree.txt`** file in a format which is generated by the command mentioned [here](https://semgrep.dev/docs/semgrep-supply-chain/setup-maven/). To bridge the gap between the SBT dep tree and the Maven format, you can take one of two possible approaches:
```

1. **Python Script Transformation**
    

>     A Python script was developed to convert the SBT `dependencyTree` output to a Maven-compatible format. The script can be run from the root of the Scala project (after setting up the `dependencyTree` SBT plugin discussed above) with the following command:

Reviewer: ```suggestion
    Use a Python script to convert the SBT `dependencyTree` output to a Maven-compatible format. The script can be run from the root of the Scala project (after setting up the `dependencyTree` SBT plugin discussed above) with the following command:
```

    
    `python3 transform.py > maven_dep_tree.txt` 
    You can find the Python script under the scala folder.
    
2. **SBT Command Transformation**
    

>     An SBT task was written to automatically transform the output and write it to a **`maven_dep_tree.txt`** file, rather than requiring a dependency on an external Python script. 

Reviewer: ```suggestion
    Use an SBT task to automatically transform the output and write it to a `maven_dep_tree.txt` file, rather than requiring a dependency on an external Python script. 
```

>     The task is called `dependencyTreeTransform` and is defined in the provided **`build.sbt`** file, which also includes some vulnerable dependencies to test if they appear in the scan. You can run `sbt dependencyTreeTransform` to execute this task.

Reviewer: ```suggestion
    The task is called `dependencyTreeTransform` and is defined in the provided `build.sbt` file. Run `sbt dependencyTreeTransform` to execute this task.
```
I checked the actual file provided here and I don't think it has the sample dependencies, which feels correct to me, thus deleting the sentence.

With either of these transformation methods, a **`maven_dep_tree.txt`** file is produced, which should be compatible with SSC scans.

### Testing the script and/or task

> Since the Python script works on intermediate output, you may receive the `dependencyTree` output from the customer rather than their original SBT file if there are issues. At the beginning of the Python script, it normally has these lines:

Reviewer: ```suggestion
Since the Python script works on intermediate output, you may want to test it with the `dependencyTree` output rather than the original SBT file if there are issues. At the beginning of the Python script, it normally has these lines:
```

```
# Run the sbt dependencyTree command and capture the output
result = subprocess.run(["sbt", "dependencyTree"], capture_output=True, text=True)

# Assign the output to actual_lines
actual_lines = result.stdout
```
But if you’re testing the intermediate output, it’s easier to have Python read the intermediate file from the command line instead:
```
result = open(sys.argv[1], 'r')
# Assign the output to actual_lines
actual_lines = result.read()
```
Then you’re good to go with testing; the rest of the file can stay the same. Just use:

`python3 transform.py {filename-of-intermediate-file}` 

> If you make any changes, you’d want to make similar changes in the script and in the task. If you don’t know Scala/Java and the changes are complicated enough that you aren’t sure how to express them, you can ask ChatGPT to help you convert your changes to a format that’s compatible with a `build.sbt` file.

Reviewer: ```suggestion
If you make any changes, you’d want to make similar changes in the script and in the task. If you have any concerns about this, feel free to reach out to us at Semgrep for assistance in integrating the changes.
```

> ### What happens if the maven_dep_tree.txt is already generated but it is not in the format SSC expect?

Reviewer: ```suggestion
### What happens if the maven_dep_tree.txt is already generated but it is not in the format SSC expects?
```
